sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -135,33 +125,47 @@ class EquivalentExpressions {
   // For some special expressions we cannot just recurse into all of its children, but we can
   // recursively add the common expressions shared between all of its children.
   private def commonChildrenToRecurse(expr: Expression): Seq[Seq[Expression]] = expr match {
+    case _: CodegenFallback => Nil
```

Comment:
not sure we can trigger this bug with some real queries, but it's an obvious bug to me.

Reply:
good catch!

Reply:
I think if we wrongly recurse into the children of `CodegenFallback`, it only produces unused subexpressions. Some redundant generated codes, i.e..

Reply:
Is it better to backport this part into branch-3.1/3.0?

Reply:
yea will do

Reply:
https://github.com/apache/spark/pull/33304